skyve-ext/src/main/java/org/skyve/impl/util/WebStatsUtil.java

```diff
@@ -2,43 +2,155 @@
+	 * previous record and sends alerts if there's a change in Ip Address and/or country
+	 * 
+	 * @param user
+	 * @param userIPAddress
+	 * @throws Exception
+	 */
+	public static void recordLogin(User user, String userIPAddress)
+			throws Exception {
 		Customer customer = user.getCustomer();
 		Module module = customer.getModule(AppConstants.ADMIN_MODULE_NAME);
-		Document loginRecordDocument = module.getDocument(customer, "UserLoginRecord");
+		Document loginRecordDocument = module.getDocument(customer, AppConstants.USER_LOGIN_RECORD_DOCUMENT_NAME);
 		AbstractPersistentBean loginRecord = loginRecordDocument.newInstance(user);
-		BindUtil.set(loginRecord, "userName", user.getName());
-		BindUtil.set(loginRecord, "loginDateTime", new DateTime(System.currentTimeMillis()));
-		BindUtil.set(loginRecord, "failed", Boolean.FALSE);
+		BindUtil.set(loginRecord, AppConstants.USER_NAME_ATTRIBUTE_NAME, user.getName());
+		BindUtil.set(loginRecord, LOGIN_DATE_TIME, new DateTime(System.currentTimeMillis()));
+		BindUtil.set(loginRecord, FAILED, Boolean.FALSE);
+		BindUtil.set(loginRecord, IP_ADDRESS, userIPAddress);
+
+		// Check if the IpInfo token has been set so as to get the country code
+		String ipInfoToken = UtilImpl.IP_INFO_TOKEN;
+		String countryCode = null;
+		if (ipInfoToken != null) {
+			final GeoIPService geoIPService = new GeoIPService();
+			Optional<String> countryCodeOptional = geoIPService.getCountryCodeForIP(userIPAddress);
+			if (countryCodeOptional.isPresent()) {
+				countryCode = countryCodeOptional.get();
+				BindUtil.set(loginRecord, COUNTRY, countryCode);
+			}
+		}
+		AbstractPersistence.get()
+				.save(loginRecordDocument, loginRecord);
+
+		// Get the last login record of the current user
+		DocumentQuery q = CORE.getPersistence()
+				.newDocumentQuery(loginRecordDocument);
+		q.getFilter()
+				.addEquals(AppConstants.USER_NAME_ATTRIBUTE_NAME, user.getName());
+		q.addBoundOrdering(LOGIN_DATE_TIME, SortDirection.descending);
+		AbstractPersistentBean previousLoginRecord = q.beanResult();
+
+		// Get email and name from the metadata user
+		Document userDocument = module.getDocument(customer, AppConstants.USER_DOCUMENT_NAME);
+		AbstractPersistentBean adminUserBean = CORE.getPersistence()
+				.retrieve(userDocument, user.getId());
+		String contactEmailBinding = BindUtil.createCompoundBinding(AppConstants.CONTACT_ATTRIBUTE_NAME,
+				AppConstants.EMAIL1_ATTRIBUTE_NAME);
+		String contactNameBinding = BindUtil.createCompoundBinding(AppConstants.CONTACT_ATTRIBUTE_NAME,
+				AppConstants.NAME_ATTRIBUTE_NAME);
+		String userEmail = (String) BindUtil.get(adminUserBean, contactEmailBinding);
+		String userName = (String) BindUtil.get(adminUserBean, contactNameBinding);
+
+		// Check if the ip address changed since last login and if so log the event
+		String lastIpAddress = (String) BindUtil.get(previousLoginRecord, IP_ADDRESS);
+		if (userIPAddress != lastIpAddress) {
```

Comment:
You need to check for nulls. lastIpAddress may be null if this is their first login?

You can use `Objects.equals` instead, but this will return true if one side is null.